File: template/cpp/template_test.cc
#include "template/cpp/template.h"
#include <gtest/gtest.h>
#include <string>
#include <iostream>
#include <sstream>

std::string captureStdout(std::function<void()> func) {
    std::stringstream buffer;
    std::streambuf* old = std::cout.rdbuf(buffer.rdbuf());

    func();

    std::cout.rdbuf(old);
    return buffer.str();
}

TEST(DownloaderTest, HTTPDownloader) {
    Downloader* downloader = new HttpDownloader();

    std::string rsp = captureStdout([&]() {
        downloader->download("http://example.com/abc.zip");
    });

    std::string expect =
        "prepare downloading\n"
        "download http://example.com/abc.zip via http,\n"
        "http save\n"
        "finish downloading";

    EXPECT_EQ(rsp, expect);

    delete downloader;
}

TEST(DownloaderTest, FTPDownloader) {
    Downloader* downloader = new FtpDownloader();

    std::string rsp = captureStdout([&]() {
        downloader->download("ftp://example.com/abc.zip");
    });

    std::string expect =
        "prepare downloading\n"
        "download ftp://example.com/abc.zip via ftp,\n"
        "default save\n"
        "finish downloading";

    EXPECT_EQ(rsp, expect);

    delete downloader;
}

File: template/template_test.go
package template

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"strings"
	"testing"
)

func captureStdout(f func()) string {
	old := os.Stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	f()

	w.Close()
	os.Stdout = old

	var buf bytes.Buffer
	io.Copy(&buf, r)

	return buf.String()
}

func TestHTTPDownloader(t *testing.T) {
	var downloader Downloader = NewHTTPDownloader()

	rsp := captureStdout(func() { downloader.Download("http://example.com/abc.zip") })

	expectRsp := strings.TrimSpace(`prepare downloading
download http://example.com/abc.zip via http,
http save
finish downloading`)
	if rsp != expectRsp {
		fmt.Println(expectRsp)
		fmt.Println(rsp)
		t.Fail()
	}
}

func TestFTPDownloader(t *testing.T) {
	var downloader Downloader = NewFTPDownloader()

	rsp := captureStdout(func() { downloader.Download("ftp://example.com/abc.zip") })
	expectRsp := strings.TrimSpace(`prepare downloading
download ftp://example.com/abc.zip via ftp,
default save
finish downloading`)

	if rsp != expectRsp {
		t.Fail()
	}
}
